### PR TITLE
Protect master and release branches

### PR DESCRIPTION
**Pull request by fejta**
/assign @geeknoid @cjwagner 

Follow-up to https://github.com/istio/test-infra/pull/1469

Martin is there a reason we closed it? This should do the same thing as in that PR with less redundant code (if we want everything defaulting unprotected, then define it that way at the top level)

### CODE REVIEW

**Comment by fejta** [comment]
Added unit tests

File: prow/config.yaml
Comment (geeknoid) [inline prow/config.yaml:70]: Why are these branches separate from the ones above? Does this have a semantic meaning?

File: prow/config.yaml
Comment (geeknoid) [inline prow/config.yaml:91]: I  think the operator also needs the same blocked branches.

File: prow/config.yaml
Comment (geeknoid) [inline prow/config.yaml:75]: We need to turn on "requires code owners review" for any istio repo other than istio/proxy and istio/istio

Comment (geeknoid) [reply prow/config.yaml:75]: I presume the behavior when something like required_status_check is not stated for a branch is that whatever is in GitHub is left alone, correct?

File: prow/config.yaml
Comment (fejta) [reply prow/config.yaml:70]: This is because istio/istio has its own complications for `release-1.1` and `release-1.2` checks

File: prow/config.yaml
Comment (fejta) [reply prow/config.yaml:75]: See the `required_pull_request_reviews` clause defined at the `istio` org level.

All repos will inherit this unless they override it. So right now all istio repos will require the `cla/google` context 
and require 1 an approving review, unless they define otherwise.


Note that [right now](https://github.com/istio/istio/settings/branch_protection_rules/1843352) istio/istio and istio/proxy both requiring these checks on any branches with protection enabled, and this PR is not changing that (are you saying it should?)

File: prow/config.yaml
Comment (fejta) [reply prow/config.yaml:91]: Done

File: prow/config.yaml
Comment (geeknoid) [reply prow/config.yaml:75]: What I'm asking for is different  than required_pull_request_reviews, we need require_code_owner_reviews: true since we use GitHub's CODEOWNERS files to control who can approve.

Comment (fejta) [reply prow/config.yaml:75]: Ack, wow those are confusingly similar settings.

Can we merge this as is -- it fixes the `release-*` branches -- and I put the `required_code_owner_reviews` changes in the next PR?

These controls exist but we basically don't use them in kubernetes so I don't have experience with them and am therefore less confident in their behavior:
* https://github.com/kubernetes/test-infra/blob/7aef384140833379d3e72f33bc0bb9d10770aefe/prow/config/branch_protection.go#L62-L71
* https://github.com/kubernetes/test-infra/blob/7aef384140833379d3e72f33bc0bb9d10770aefe/prow/config/branch_protection.go#L38-L39

Let me figure out how to express that to the tool and I'll send a PR

So long as we leave `required_code_owner_reviews` as `nil` the tool should not alter these settings where they already exist.

**Comment by fejta** [comment]
@geeknoid can you try again with an approving review instead of the slash commands? and/or would you want to dig back through the logs to see whether we dropped that webhook?

**Review by howardjohn — APPROVED** [review]
possibly related to missing that: https://github.com/istio/test-infra/issues/1501 and https://github.com/istio/test-infra/issues/1500

**Comment by fejta** [comment]
Reopening #1452 to see if we can get traction on this. We do not ever expect prow to drop webhooks

**Comment by fejta** [comment]
hello

/test lint-check

world

**Comment by fejta** [comment]
That comment worked :-/ 

I wonder what the deal is...

**Comment by geeknoid** [comment]
I  don't know how this behaves at runtime, but the YAML appears invalid. Line 70 reuses the same key (blocked_branches)  which  my  editor is highlighting as invalid.

**Comment by geeknoid** [comment]
Also, the operator repo doesn't have  branch protection for the release-1.1 branch. Perhaps it's related to the YAML issue from above?

**Comment by fejta** [comment]
This is part of why I wrote the unit test to prove that it was working correctly. I find yaml anchors kind of not worth the effort (although probably necessary in this instance so long as we require random branches to remain unprotected).

So I'm pretty sure it is valid yaml -- prow parses it correctly and the concrete values match what we want thus far. Let me see if I can repo the release-1.1 branch issue.

**Comment by fejta** [comment]
https://github.com/istio/test-infra/pull/1509 proves things are configured correctly (at least at the yaml level) as best as I can tell.

Any chance your editor has a yaml parsing bug?

**Comment by cjwagner** [comment]
FYI you can see if the config is being parsed as you expect by checking the parsed config at https://prow.istio.io/config

**Comment by geeknoid** [comment]
Quite possible my editor doesn't understand the anchor syntax.

Any idea why the operator repo doesn't have branch protection for its release-1.1 branch?

**Comment by fejta** [comment]
Yes, https://github.com/istio/test-infra/issues/1511 I think

**Pull request merged**
